llpc/tool/llpcPipelineBuilder.h

```diff
@@ -30,49 +30,84 @@
 public:
-  // Creates a PipelineBuilder.
+  // Initializes PipelineBuilder. Use `createPipelineBuilder` to create concrete instances of this class.
   //
   // @param compiler : LLPC compiler object.
   // @param [in/out] compileInfo : Compilation info of LLPC standalone tool. This will be modified by `build()`.
-  // @param dumpOptions : Pipeline dump options. Pipeline dumps are disabled when llvm::None is passed.
+  // @param dumpOptions : Pipeline dump options. Pipeline dumps are disabled when `llvm::None` is passed.
   // @param printPipelineInfo : Whether to print pipeline info (hash, filenames) before compilation.
   PipelineBuilder(ICompiler &compiler, CompileInfo &compileInfo, llvm::Optional<Vkgc::PipelineDumpOptions> dumpOptions,
                   bool printPipelineInfo)
       : m_compiler(compiler), m_compileInfo(compileInfo), m_dumpOptions(std::move(dumpOptions)),
         m_printPipelineInfo(printPipelineInfo) {}
 
+  virtual ~PipelineBuilder() = default;
+
   // Compiles the pipeline and performs linking.
-  LLPC_NODISCARD Result build();
+  // The implementations should call `runPreBuildActions` before performing compilation with `m_compiler` and
+  // call `runPostBuildActions` after.
+  //
+  // @returns : Calculated pipeline hash.
+  virtual llvm::Error build() = 0;
 
-private:
-  // Returns true iff pipeline dumps are requested.
-  LLPC_NODISCARD bool shouldDumpPipelines() const { return m_dumpOptions.hasValue(); }
+  // Calculates the hash of the compiled pipeline. This is used by `printPipelineInfo` to produce verbose logs.
+  //
+  // @param buildInfo : Pipeline build information.
+  // @returns : Calculated pipeline hash.
+  LLPC_NODISCARD virtual uint64_t getPipelineHash(Vkgc::PipelineBuildInfo buildInfo) = 0;
 
-  // Builds graphics pipeline and does linking.
-  LLPC_NODISCARD Result buildGraphicsPipeline(Vkgc::BinaryData &outPipeline);
+  // Returns the compiler.
+  //
+  // @returns : Compiler handle.
+  ICompiler &getCompiler() { return m_compiler; }
 
-  // Builds compute pipeline and does linking.
-  LLPC_NODISCARD Result buildComputePipeline(Vkgc::BinaryData &outPipeline);
+  // Returns the compile info.
+  //
+  // @returns : Compile info.
+  CompileInfo &getCompileInfo() { return m_compileInfo; }
+
+  // Returns the pipeline dump options.
+  //
+  // @returns : `PipelineDumpOptions` or `llpc::None` if pipeline dumps were not requested.
+  llvm::Optional<Vkgc::PipelineDumpOptions> &getDumpOptions() { return m_dumpOptions; }
+
+  // Returns true iff pipeline dumps are requested.
+  //
+  // @returns : `true` is pipeline dumps were requested, `false` if not.
+  LLPC_NODISCARD virtual bool shouldDumpPipelines() const { return m_dumpOptions.hasValue(); }
 
   // Runs optional pre-build code (pipeline dumping, pipeline info printing).
-  LLPC_NODISCARD void *runPreBuildActions(Vkgc::PipelineBuildInfo buildInfo);
+  LLPC_NODISCARD virtual void *runPreBuildActions(Vkgc::PipelineBuildInfo buildInfo);
```

Comment:
It seems that `runPreBuildActions` and `runPostBuildActions` are called manually by each implementation, meaning a new implementation of `PipelineBuilder` that overrides these two methods is not going to see them called unless they call them manually in their `build` implementation. Given that, is there a reason in having them as base class methods at all, or we should rather have them as methods implemented by each derived class, if needed?

Reply:
Good point. I considered a few options:
*  Making `build()` a static function and forcing the order of operations onto the implementations. I do not like this approach because it will force all derived classes to move function arguments to class members, making the code more difficult to understand.
*  Removing them  from the base builder and adding a common intermediate class for graphics and compute pipelines. I do not like this approach because it effective leaves the base class empty and not very useful.
*  Dropping `virtual` from some of the member functions. This should be fine because `build` is virtual anyway, so derived classes can treat `runPre/PostBuildAction` as helper functions, or not use them at all if they want to do something completely different.

I went ahead with the last one, as it does not hurt code readability while providing enough implementation freedom.